### PR TITLE
expose resources through resource() function in world API

### PR DESCRIPTION
**Pull request by remybar**
See https://github.com/dojoengine/dojo/issues/2244.

## Summary by CodeRabbit

- **New Features**
	- Introduced a new `Resource` enum to enhance resource management and accessibility in the world module.
	- Added a consolidated `resource` function to replace the previous `model` and `contract` functions, simplifying resource retrieval.
	- Improved error handling in resource retrieval methods to enhance robustness and type safety.
	- Automation of Cairo project builds with the new `build_cairo_projects.sh` script.

- **Bug Fixes**
	- Enhanced test coverage to validate resource types and improve overall reliability.

- **Documentation**
	- Updated manifest files to reflect new resource definitions and modifications to class hashes.

- **Chores**
	- Modified CI workflow to improve environment setup and script execution for builds.
	- Streamlined build scripts for development and release configurations.

### CODE REVIEW

File: crates/dojo-core/src/world/world_contract.cairo [hunk 1]
```diff
@@ -734,12 +709,17 @@ pub mod world {
...
+
+            if let Resource::Contract((_, contract_address)) = self.resources.read(selector) {
+                IUpgradeableDispatcher { contract_address }.upgrade(class_hash);
+                EventEmitter::emit(
+                    ref self, ContractUpgraded { class_hash, address: contract_address }
+                );
+                class_hash
+            } else {
+                core::panic_with_felt252(Errors::INVALID_RESOURCE_SELECTOR);
+                core::num::traits::Zero::<ClassHash>::zero()
```
Comment (glihm) [inline crates/dojo-core/src/world/world_contract.cairo:721]: We don't need to return anything if we panic. If you remove the `;` after the panic the compiler should not complain of different types in the `if let` branch.

File: crates/dojo-core/src/world/world_contract.cairo [hunk 2]
```diff
@@ -6,12 +6,20 @@ use dojo::model::{ModelIndex, ResourceMetadata};
...
 use dojo::utils::bytearray_hash;
 
+#[derive(Drop, starknet::Store, Serde, Default, Debug)]
+pub enum Resource {
+    Model: (ClassHash, ContractAddress),
+    Contract: (ClassHash, ContractAddress),
+    Namespace,
+    World,
+    #[default]
+    None,
```
Comment (glihm) [inline crates/dojo-core/src/world/world_contract.cairo:16]: Wondering if `None` is ok, or if we should change for `Unknown`/`Unregistered`. Very open question, no strict preference on my end. Was wondering what you guys think.

**Review by glihm — COMMENTED** [review]
Thanks for the rework @remybar, some minor comment and let's merge!

File: crates/dojo-core/src/world/world_contract.cairo (hunk 1 above)
Comment (remybar) [reply crates/dojo-core/src/world/world_contract.cairo:721]: Ha ! That's the trick ! Thanks 👍

File: crates/dojo-core/src/world/world_contract.cairo (hunk 2 above)
Comment (remybar) [reply crates/dojo-core/src/world/world_contract.cairo:16]: Yes, why not ! Why one do you prefer ? Maybe `Unregistered` is the most explicit/specific term ?

Comment (glihm) [reply crates/dojo-core/src/world/world_contract.cairo:16]: Yup let's go with `Unregistered`. 👍

**Review by glihm — APPROVED** [review]
Thanks for that @remybar. 🙏

**Pull request merged**